### PR TITLE
791 manual id entry

### LINKED ISSUE #791: Programs enrolment IDs / manual entered IDs
- [x] the patient "Program ID" is the unique identifier given to a patient for a specific program. Currently named "Patient enrolment ID"
- [x] System should generate a sequential Program ID (and it is OK if there are gaps in the sequence) but also have the ability to enter a Program ID manually (we should avoid duplicate though).

### PR DESCRIPTION
**Pull request by clemens-msupply**
Includes:
- IdGenerator option to manually update an id
- validation check do avoid duplicated ids (only support for program enrollment patient ids)
- same duplicate check when auto generating an id. In case of a conflict auto generate is continued to be called till a unique id is found. (possible problem: in case of many manual id entries, the autogenerate might need a long time to catch up with the latest manual entered id?)
- Refactor out an DebouncedTextInput from the Text control (needed for the manual entry)
- Refactor out and fix custom JSONForms errors from the Text control
- Move program enrollment queries to the programs package

Closes #791

### CODE REVIEW

**Comment by CarlosNZ** [comment]
Looks pretty good. One regression I noticed: when you add a program, it no longer immediately shows up in the "Programs" list for that patient, I needed to either refresh the page or click another link and go back in.

So somehow, it's no longer re-querying the "Programs" list for that patient on submission of a new one.

**Comment by clemens-msupply** [comment]
> Looks pretty good. One regression I noticed: when you add a program, it no longer immediately shows up in the "Programs" list for that patient, I needed to either refresh the page or click another link and go back in.
> 
> So somehow, it's no longer re-querying the "Programs" list for that patient on submission of a new one.

Thanks! There are currently two program enrollment apis that are going to be merged. They where using different base keys. I fixed it.

File: client/packages/programs/src/JsonForms/common/components/DebouncedTextInput.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  return (
+    <BasicTextInput
+      {...inputProps}
+      value={localData ?? ''}
+      onChange={e => {
+        setLatestKey(Date.now());
+        setLocalData(e.target.value);
+        onChangeDebounced(e.target.value);
+      }}
+    />
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/common/components/DebouncedTextInput.tsx:67]: Yeah, as we discussed, might be good to make this just a `useDebouncedInput` hook and keep the presentational components as they are, but I can do this as part of #1015.

**Review by CarlosNZ — APPROVED** [review]
Okay, looks good. Nice refactoring. Prob best to merge this and I'll keep an eye out for any additional problems when I do the next tidy-up in #1015

**Comment by clemens-msupply** [comment]
thanks for the review!

**Pull request merged**